File: example_code/Optimnization&Vectorization/Exercise/C/MatMul1.c
#include <stdio.h>
#include <time.h>
#include <stdlib.h>
#include <math.h>
#define N 1024

double WallTime()
{
    struct timespec tp;
    clock_gettime(CLOCK_REALTIME,&tp);
    return ((double)tp.tv_sec + (double)tp.tv_nsec * 1.e-9);
}

int main(void)
{
  int i,j,k;
  double c0,c1,c2,c3,tmp;
  double a[N][N],b[N][N],c[N][N];
  double stime, etime;
  double dtime;
 
  for(i=0;i<N;i++){
    for(j=0;j<N;j++){
      a[i][j]=(double)(rand())/RAND_MAX;
      b[i][j]=(double)(rand())/RAND_MAX;
      c[i][j]=0.0;
    }
  }

  stime=WallTime();
  for(i=0;i<N;i++){
    for(j=0;j<N;j++){
      for(k=0;k<N;k++){
        c[i][j]=c[i][j]+a[i][k]*b[k][j];
      }
    }
  }
  etime=WallTime();
  dtime=etime-stime;
  printf("%f\n",c[1][1]);
  printf("Original code(No Opt) : %12.5f\n",dtime);

  for(i=0;i<N;i++){
    for(j=0;j<N;j++){
      c[i][j]=0.0;
    }
  }

  stime=WallTime();
// TODO
  etime=WallTime();
  dtime=etime-stime;
  printf("%f\n",c[1][1]);
  printf("Unrolling(Opt) : %12.5f\n",dtime);
  return 0;
}

File: example_code/Optimnization&Vectorization/Exercise/Fortran/MatMul1.f90
PROGRAM MATMUL1
IMPLICIT NONE
INTEGER,PARAMETER::N=1024
INTEGER::I,J,K
INTEGER(KIND=8)::ST,ET,C_RATE
REAL(KIND=8)::DTIME
REAL(KIND=8),DIMENSION(N,N)::A,B,C
REAL(KIND=8)::C0,C1,C2,C3,TMP

CALL RANDOM_NUMBER(A)
CALL RANDOM_NUMBER(B)
C=0.0

CALL SYSTEM_CLOCK(ST,C_RATE)
DO J=1,N
  DO I=1,N
    DO K=1,N
      C(I,J)=C(I,J)+A(I,K)*B(K,J)
    END DO
  END DO
END DO
CALL SYSTEM_CLOCK(ET)
DTIME=(ET-ST)/DBLE(C_RATE)
PRINT*,C(1,1)
PRINT*,'Original code(No Opt):',DTIME

C=0.0
CALL SYSTEM_CLOCK(ST,C_RATE)
!! TODO
CALL SYSTEM_CLOCK(ET)
DTIME=(ET-ST)/DBLE(C_RATE)
PRINT*,C(1,1)
PRINT*,'Unrolling(Opt):',DTIME
END PROGRAM MATMUL1
